crates/handlers/src/views/recovery/finish.rs

```diff
@@ -0,0 +1,246 @@
+// Copyright 2024 The Matrix.org Foundation C.I.C.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use anyhow::Context;
+use axum::{
+    extract::{Query, State},
+    response::{Html, IntoResponse, Response},
+    Form,
+};
+use mas_axum_utils::{
+    cookies::CookieJar,
+    csrf::{CsrfExt, ProtectedForm},
+    FancyError,
+};
+use mas_policy::Policy;
+use mas_router::UrlBuilder;
+use mas_storage::{BoxClock, BoxRepository, BoxRng};
+use mas_templates::{
+    ErrorContext, FieldError, FormState, RecoveryFinishContext, RecoveryFinishFormField,
+    TemplateContext, Templates,
+};
+use serde::{Deserialize, Serialize};
+use zeroize::Zeroizing;
+
+use crate::{passwords::PasswordManager, PreferredLanguage};
+
+#[derive(Deserialize)]
+pub(crate) struct RouteQuery {
+    ticket: String,
+}
+
+#[derive(Deserialize, Serialize)]
+pub(crate) struct RouteForm {
+    new_password: String,
+    new_password_confirm: String,
+}
+
+pub(crate) async fn get(
+    mut rng: BoxRng,
+    clock: BoxClock,
+    mut repo: BoxRepository,
+    State(templates): State<Templates>,
+    PreferredLanguage(locale): PreferredLanguage,
+    cookie_jar: CookieJar,
+    Query(query): Query<RouteQuery>,
+) -> Result<Response, FancyError> {
+    let (csrf_token, cookie_jar) = cookie_jar.csrf_token(&clock, &mut rng);
+
+    let ticket = repo
+        .user_recovery()
+        .find_ticket(&query.ticket)
+        .await?
+        .context("Unknown ticket")?;
+
+    let session = repo
+        .user_recovery()
+        .lookup_session(ticket.user_recovery_session_id)
+        .await?
+        .context("Unknown session")?;
```

Comment:
This should probably say something like it was invalid or expired etc, rather than 500ing

Reply:
Later on I added the 'consumed at' check separately, so that exact error would only ever happen if the foreign key is broken for some reason?

Reply:
I think I mean for the 'Unknown ticket' in this case then

Reply:
That wouldn't be the first place we have this. As we're not deleting tickets for now, it's probably fine, and then it's a case of #2880?